File: Go/Data Structure/Array/dynamicArray.go
package main

import (
	"fmt"
)

type DynamicArray struct {
	size int
	arr  []string
	len  int
}

func NewDynamicArray(arraySize int) *DynamicArray {
	return &DynamicArray{size: arraySize, arr: make([]string, arraySize), len: 0}
}

func (arr *DynamicArray) Push(item string) {
	if arr.len == arr.size {
		arr.Resize()
	}
	arr.arr[arr.len] = item
	arr.len++
	return
}

func (arr *DynamicArray) Insert(index int, item string) {
	if index < 0 || index > arr.len {
		fmt.Println("Index out of Range")
		return
	}

	for i := arr.len; i >= index; i-- {
		arr.arr[i] = arr.arr[i-1]
	}
	arr.arr[index] = item
	arr.len++
	return
}

func (arr *DynamicArray) Delete(index int) {
	if index < 0 || index > arr.len {
		fmt.Println("Index out of Range")
		return
	}

	for i := index; i < arr.len-1; i++ {
		arr.arr[i] = arr.arr[i+1]
	}
	arr.arr[arr.len-1] = ""
	arr.len--
	return
}

func (arr *DynamicArray) GetLen() int {
	return arr.len
}

func (arr *DynamicArray) Search(item string) int {
	for i := 0; i < arr.len; i++ {
		if arr.arr[i] == item {
			return i
		}
	}
	return -1
}

func (arr *DynamicArray) Resize() {
	newSize := arr.size * 2
	newArray := make([]string, newSize)

	for i := 0; i < len(arr.arr); i++ {
		newArray[i] = arr.arr[i]
	}
	arr.arr = newArray
	arr.size = newSize
}

func (arr *DynamicArray) Shrink() {
	if arr.len < arr.size/2 {
		newSize := int(arr.size / 2)
		newArray := make([]string, newSize)

		for i := range arr.len {
			newArray[i] = arr.arr[i]
		}
		arr.arr = newArray
		arr.size = newSize
	}
}

func main() {
	array := NewDynamicArray(10)

	array.Push("Assembly")
	array.Push("A.Net")
	array.Push("ALGOL")
	array.Push("Ada")
	array.Push("BASIC")
	array.Push("BCPL")
	array.Push("C")
	array.Push("C#")
	array.Push("COBOL")
	array.Push("Chuck")
	array.Push("Delphi")

	fmt.Println(array.arr)

	array.Delete(5)

	fmt.Println(array.arr)

	array.Insert(9, "Machine Code")

	array.Delete(9)
	array.Delete(8)
	array.Delete(7)
	array.Delete(6)

	fmt.Println(array.arr, array.size, array.len)

	array.Shrink()

	fmt.Println(array.arr, array.size, array.len)
}

File: Nim/Data Structure/Array/dynamic_array.nim
type
    DynamicArray = ref object
        capacity: int
        data: seq[int]
        length: int

proc new_array(capacity: int): DynamicArray = 
    DynamicArray(capacity: capacity, data: newSeq[int](capacity), length: 0)


proc resize(self: var DynamicArray) = 
    let new_size = 2 * self.capacity
    var new_array = newSeq[int](new_size)

    for i in 0..<self.data.len: 
        new_array[i] = self.data[i]
    
    self.data = new_array
    self.capacity = new_size

proc push(self: var DynamicArray, item: int) = 
    if self.length == self.capacity:
        self.resize()
    
    self.data[self.length] = item
    self.length += 1

proc insert(self: var DynamicArray, index: int, item: int) =
  if index < 0 or index > self.length:
    echo "Index out of Range"
    return
  
  if self.length == self.capacity:
    self.resize()
  
  for i in countdown(self.length, index + 1):
    self.data[i] = self.data[i-1]
  
  self.data[index] = item
  self.length += 1

proc delete(self: var DynamicArray, index: int) = 
    if index < 0 or index > self.length:
        echo "Index out of range"
        return
    
    for i in index..<self.length-1:
        self.data[i] = self.data[i+1]
    
    self.data[self.length-1] = 0
    self.length -= 1

proc shrink(self: var DynamicArray) = 
    if self.length < self.capacity div 2:
        let new_size = self.capacity div 2
        var new_array = newSeq[int](new_size)

        for i in 0 ..< self.length:
            new_array[i] = self.data[i]
        
        self.capacity = new_size
        self.data = new_array

proc search(self: var DynamicArray, item: int): int =
    for i in 0 ..< self.length:
        if self.data[i] == item:
            return i
    return -1

proc get_length(self: var DynamicArray): int = 
    return self.length

proc toString(self: var DynamicArray): string = 
    result = $self.data